src/execution_tree/primitives/dictionary.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2017-2018 Weile Wei
+// Copyright (c) 2018 Parsa Amini
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/ir/dictionary.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <phylanx/execution_tree/primitives/primitive_argument_type.hpp>
+#include <phylanx/util/variant.hpp>
+#include <boost/functional/hash.hpp>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives {
+
+}}}
+
+namespace std {
+    std::size_t hash<phylanx::util::recursive_wrapper<
+        phylanx::execution_tree::primitive_argument_type>>::
+    operator()(argument_type const& s) const noexcept
+    {
+        phylanx::execution_tree::primitive_argument_type const& val = s.get();
+        std::size_t result;
+        switch (val.index())
+        {
+        case 0:    // ast::nil
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "phylanx::ir::dictionary",
+                "unsupported nil data type");
+
+        case 1:    // phylanx::ir::node_data<std::uint8_t>
+        {
+            phylanx::ir::node_data<std::uint8_t> val_temp =
+                phylanx::util::get<1>(val);
+            switch (val_temp.num_dimensions())
```

Comment:
This creates a copy of the `node_data` object. It might not be worth avoiding, but you can try ;-) (same is true below for `node_data<std::int64_t>` and `node_data<double>`).